src/platform/datapath_raw_xdp.c

```diff
@@ -221,232 +232,252 @@ CxPlatDpRawInitialize(
     const uint32_t RxPacketSize = ALIGN_UP(RxHeadroom + MAX_ETH_FRAME_SIZE, XDP_RX_PACKET);
     QUIC_STATUS Status;
 
-    InitializeSListHead(&Xdp->RxPool);
-    InitializeSListHead(&Xdp->TxPool);
-    CxPlatLockInitialize(&Xdp->TxLock);
-    CxPlatListInitializeHead(&Xdp->TxQueue);
-    CxPlatListInitializeHead(&Xdp->WorkerTxQueue);
-    Xdp->WorkerRxPool.Next = NULL;
-
     CxPlatXdpReadConfig(Xdp);
     Datapath->Cpu = Xdp->DatapathCpuNumber;
     CxPlatDpRawGenerateCpuTable(Datapath);
 
-    //
-    // RX datapath.
-    //
-
-    Xdp->RxBuffers = CxPlatAlloc(Xdp->RxBufferCount * RxPacketSize, RX_BUFFER_TAG);
-    if (Xdp->RxBuffers == NULL) {
+    Xdp->Queues = CxPlatAlloc(Xdp->QueueCount * sizeof(*Xdp->Queues), QUEUE_TAG);
+    if (Xdp->Queues == NULL) {
         QuicTraceEvent(
             AllocFailure,
             "Allocation of '%s' failed. (%llu bytes)",
-            "XDP RX Buffers",
-            Xdp->RxBufferCount * RxPacketSize);
+            "XDP Queues",
+            Xdp->QueueCount * sizeof(*Xdp->Queues));
         Status = QUIC_STATUS_OUT_OF_MEMORY;
         goto Error;
     }
 
-    Status = XskCreate(&Xdp->RxXsk);
-    if (QUIC_FAILED(Status)) {
-        QuicTraceEvent(
-            LibraryErrorStatus,
-            "[ lib] ERROR, %u, %s.",
-            Status,
-            "XskCreate");
-        goto Error;
-    }
+    CxPlatZeroMemory(Xdp->Queues, Xdp->QueueCount * sizeof(*Xdp->Queues));
 
-    XSK_UMEM_REG RxUmem = {0};
-    RxUmem.address = Xdp->RxBuffers;
-    RxUmem.chunkSize = RxPacketSize;
-    RxUmem.headroom = RxHeadroom;
-    RxUmem.totalSize = Xdp->RxBufferCount * RxPacketSize;
+    for (uint32_t QueueIndex = 0; QueueIndex < Xdp->QueueCount; QueueIndex++) {
+        XDP_QUEUE* Queue = &Xdp->Queues[QueueIndex];
 
-    Status = XskSetSockopt(Xdp->RxXsk, XSK_SOCKOPT_UMEM_REG, &RxUmem, sizeof(RxUmem));
-    if (QUIC_FAILED(Status)) {
-        QuicTraceEvent(
-            LibraryErrorStatus,
-            "[ lib] ERROR, %u, %s.",
-            Status,
-            "XskSetSockopt(XSK_SOCKOPT_UMEM_REG)");
-        goto Error;
-    }
+        InitializeSListHead(&Queue->RxPool);
+        InitializeSListHead(&Queue->TxPool);
+        CxPlatLockInitialize(&Queue->TxLock);
+        CxPlatListInitializeHead(&Queue->TxQueue);
+        CxPlatListInitializeHead(&Queue->WorkerTxQueue);
+        CxPlatZeroMemory(&Queue->WorkerRxPool, sizeof(Queue->WorkerRxPool));
```

Comment:
Wasn't this already zero'ed with everything else above?

Reply:
Yeah, I was debating this - previously I had been explicitly initializing the Next field to null. I'll just remove the init.